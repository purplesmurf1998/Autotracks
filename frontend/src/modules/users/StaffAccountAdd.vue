<template>
  <div>
    <CForm @submit.prevent="submitForm">
      <CAlert v-if="!!errorMessage" color="danger">{{ errorMessage }}</CAlert>
      <CRow>
        <CCol>
          <CInput
            label="First Name"
            placeholder="User first name"
            :value.sync="first_name"
          />
        </CCol>
        <CCol>
          <CInput
            label="Last Name"
            placeholder="User last name"
            :value.sync="last_name"
          />
        </CCol>
      </CRow>
      <CInput
        label="Email"
        placeholder="User email address"
        :value.sync="email"
      />
      <CSelect
        label="Account Role"
        id="role-select"
        :options="roles"
        :value.sync="role"
        placeholder="Please select a role"
      />
      <CRow>
        <CCol>
          <label>Account Password</label>
        </CCol>
        <CCol class="d-flex align-items-end flex-column">
          <p style="color: red">{{ password }}</p>
        </CCol>
      </CRow>
      <label
        >Take note of the password and pass on to staff member. User will be
        prompted to change password upon first login.</label
      >

      <CRow class="justify-content-center mt-2">
        <CButton
          color="primary"
          type="submit"
          :disabled="disableButtons"
          id="create-staff"
        >
          Create
        </CButton>
        <CButton
          class="ml-1"
          color="secondary"
          :disabled="disableButtons"
          @click="setAddingStaffAccount(false)"
        >
          Cancel
        </CButton>
      </CRow>
    </CForm>
  </div>
</template>

<script>
const axios = require("axios");

export default {
  name: "StaffAccountAdd",
  props: ["setAddingStaffAccount", "addNewStaffAccount"],
  data() {
    return {
      disableButtons: false,
      errorMessage: null,
      roles: [
        "Management",
        "Sales Rep",
        "Reception",
      ],
      first_name: "",
      last_name: "",
      email: "",
      role: "",
      password: this.generateRandomPassword(),
      permissions: [],
    };
  },
  methods: {
    generateRandomPassword() {
      /*
      Taken from stackoverflow post: https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
      */
      var result = "";
      var characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ123456789";
      var charactersLength = characters.length;
      for (var i = 0; i < 12; i++) {
        result += characters.charAt(
          Math.floor(Math.random() * charactersLength)
        );
      }
      return result;
    },
    formIsValid() {
      return (
        this.first_name != "" &&
        this.last_name != "" &&
        this.email != "" &&
        this.role != ""
      );
    },
    showErrorMessage(message) {
      this.errorMessage = message;
      setTimeout(() => {
        this.errorMessage = null;
      }, 5000);
    },
    submitForm(form) {

      // create the data to be passed
      const data = {
        first_name: this.first_name,
        last_name: this.last_name,
        email: this.email,
        role: this.role,
        dealership: this.$route.params.dealershipId,
        password: this.password,
      };

      // send post request to the API
      if (this.formIsValid()) {
        axios({
          method: "POST",
          url: `${this.$store.state.api}/users`,
          headers: {
            Authorization: `Bearer ${this.$store.state.auth.token}`,
          },
          data,
        })
          .then((response) => {
            if (response.data.success) {
              this.$store.commit("setProperty", ["createUserCompleted", true]);
              this.addNewStaffAccount(response.data.payload);
              this.setAddingStaffAccount(false);
            }
          })
          .catch((err) => {
            console.log(err);
            this.showErrorMessage("Failed to create staff account.");
          });
      } else {
        this.showErrorMessage("Invalid form. Please verify the field entries.");
      }
    },
  },
};
</script>
-
<style>
</style>
