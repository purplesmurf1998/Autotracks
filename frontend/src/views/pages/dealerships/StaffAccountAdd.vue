<template>
  <div>
    <CForm @submit.prevent="submitForm">
      <CAlert v-if="!!errorMessage" color="danger">{{ errorMessage }}</CAlert>
      <CRow>
        <CCol>
          <CInput
            label="First Name"
            placeholder="User first name"
            :value.sync="first_name"
          />
        </CCol>
        <CCol>
          <CInput
            label="Last Name"
            placeholder="User last name"
            :value.sync="last_name"
          />
        </CCol>
      </CRow>
      <CInput
        label="Email"
        placeholder="User email address"
        :value.sync="email"
      />
      <CSelect
        label="Account Role"
        id="role-select"
        :options="roles"
        :value.sync="role"
        @change="setDefaultSelected"
        placeholder="Please select a role"
      />
      <label class="mb-0">Account permissions</label>
      <hr />
      <CRow>
        <CCol>
          <CRow v-for="(permission, index) in permissionsListLeft" :key="index">
            <CCol>
              <label>{{ permission }}</label>
            </CCol>
            <CCol class="d-flex justify-content-end">
              <CSwitch
                :id="permission"
                class="mr-1"
                color="success"
                :checked="isPermissionSelected(permission)"
                @change.native="togglePermissionChange"
                shape="pill"
              />
            </CCol>
          </CRow>
        </CCol>
        <CCol>
          <CRow
            v-for="(permission, index) in permissionsListRight"
            :key="index"
          >
            <CCol>
              <label>{{ permission }}</label>
            </CCol>
            <CCol class="d-flex justify-content-end">
              <CSwitch
                :id="permission"
                class="mr-1"
                color="success"
                :checked="isPermissionSelected(permission)"
                @change.native="togglePermissionChange"
                shape="pill"
              />
            </CCol>
          </CRow>
        </CCol>
      </CRow>
      <hr />
      <CRow>
        <CCol>
          <label>Account Password</label>
        </CCol>
        <CCol class="d-flex align-items-end flex-column">
          <p style="color: red">{{ password }}</p>
        </CCol>
      </CRow>
      <label
        >Take note of the password and pass on to staff member. User will be
        prompted to change password upon first login.</label
      >

      <CRow class="justify-content-center mt-2">
        <CButton
          color="primary"
          type="submit"
          :disabled="disableButtons"
          id="create-staff"
        >
          Create
        </CButton>
        <CButton
          class="ml-1"
          color="danger"
          :disabled="disableButtons"
          @click="setAddingStaffAccount(false)"
        >
          Cancel
        </CButton>
      </CRow>
    </CForm>
  </div>
</template>

<script>
const axios = require("axios");

export default {
  name: "StaffAccountAdd",
  props: ["setAddingStaffAccount", "addNewStaffAccount"],
  data() {
    return {
      disableButtons: false,
      errorMessage: null,
      roles: [
        "Management",
        "Sales Rep",
        "Sales Rep + Showroom",
        "Sales Rep + Demoline",
        "Sales Rep + Benefits",
        "Reception",
      ],
      permissionsListLeft: [
        "Add Dealerships",
        "View Dealerships",
        "Edit Dealerships",
        "Delete Dealerships",
        "Add Staff Users",
        "View Staff Users",
        "Edit Staff Users",
      ],
      permissionsListRight: [
        "Delete Staff Users",
        "Add Vehicles",
        "View Vehicles",
        "Edit Vehicle Properties",
        "Edit Vehicle Locations",
        "Sell Vehicles",
        "Delete Vehicles",
      ],
      first_name: "",
      last_name: "",
      email: "",
      role: "",
      password: this.generateRandomPassword(),
      permissions: [],
    };
  },
  methods: {
    generateRandomPassword() {
      /*
      Taken from stackoverflow post: https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
      */
      var result = "";
      var characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ123456789";
      var charactersLength = characters.length;
      for (var i = 0; i < 12; i++) {
        result += characters.charAt(
          Math.floor(Math.random() * charactersLength)
        );
      }
      return result;
    },
    isPermissionSelected(permission) {
      const index = this.permissions.indexOf(permission);
      return index >= 0;
    },
    togglePermissionChange(element) {
      const checked = element.target.checked;
      const permission = element.target.id;

      if (checked) {
        // add the permission to the list
        this.addPermission(permission);
      } else {
        this.removePermission(permission);
      }
    },
    addPermission(permission) {
      this.permissions.push(permission);
    },
    removePermission(permission) {
      const index = this.permissions.indexOf(permission);
      this.permissions.splice(index, 1);
    },
    setDefaultSelected() {
      this.permissions = [];
      this.permissions = this.getDefaultSelected();
    },
    getDefaultSelected() {
      switch (this.role) {
        case "Management":
          return [
            "View Dealerships",
            "Add Staff Users",
            "View Staff Users",
            "Edit Staff Users",
            "Delete Staff Users",
            "View Vehicles",
            "Edit Vehicle Locations",
          ];
        case "Sales Rep":
          return ["View Vehicles", "Edit Vehicle Locations", "Sell Vehicles"];
        case "Sales Rep + Showroom":
          return ["View Vehicles", "Edit Vehicle Locations", "Sell Vehicles"];
        case "Sales Rep + Demoline":
          return ["View Vehicles", "Edit Vehicle Locations", "Sell Vehicles"];
        case "Sales Rep + Benefits":
          return ["View Vehicles", "Edit Vehicle Locations", "Sell Vehicles"];
        case "Reception":
          return [
            "Add Vehicles",
            "View Vehicles",
            "Edit Vehicle Properties",
            "Edit Vehicle Locations",
            "Delete Vehicles",
          ];
        default:
          return ["View Vehicles"];
      }
    },
    formIsValid() {
      return (
        this.first_name != "" &&
        this.last_name != "" &&
        this.email != "" &&
        this.role != ""
      );
    },
    showErrorMessage(message) {
      this.errorMessage = message;
      setTimeout(() => {
        this.errorMessage = null;
      }, 5000);
    },
    submitForm(form) {
      // console.log("####### STAFF USER FORM SUBMIT ########");
      // console.log(`First Name: ${this.first_name}`);
      // console.log(`Last Name: ${this.last_name}`);
      // console.log(`Email: ${this.email}`);
      // console.log(`Role: ${this.role}`);
      // console.log(`Permissions: ${this.selected}`);
      // console.log(form.target.permissions.selected);

      // create the data to be passed
      const data = {
        first_name: this.first_name,
        last_name: this.last_name,
        email: this.email,
        role: this.role,
        dealership: this.$route.params.dealershipId,
        permissions: this.permissions,
        password: this.password,
      };

      // send post request to the API
      if (this.formIsValid()) {
        axios({
          method: "POST",
          url: "http://localhost:5000/api/v1/users",
          headers: {
            Authorization: `Bearer ${this.$store.state.auth.token}`,
          },
          data,
        })
          .then((response) => {
            if (response.data.success) {
              this.$store.commit("setProperty", ["createUserCompleted", true]);
              this.addNewStaffAccount(response.data.payload);
              this.setAddingStaffAccount(false);
            }
          })
          .catch((err) => {
            console.log(err);
            this.showErrorMessage("Failed to create staff account.");
          });
      } else {
        this.showErrorMessage("Invalid form. Please verify the field entries.");
      }
    },
  },
};
</script>

<style>
</style>