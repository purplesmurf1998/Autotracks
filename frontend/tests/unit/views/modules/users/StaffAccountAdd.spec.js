import {createLocalVue, shallowMount} from '@vue/test-utils'
import CoreuiVue from '@coreui/vue-pro'
import Vuex from 'vuex'

import StaffAccountAdd from './../../../../../src/modules/users/StaffAccountAdd.vue'

const localVue = createLocalVue()
localVue.use(CoreuiVue)
localVue.use(Vuex)

const store = new Vuex.Store({
    state: {
      sidebarShow: 'responsive',
      sidebarMinimize: false,
      asideShow: false,
      darkMode: false
    }
  })

  describe('StaffAccountAdd.vue', () => {
    const wrapper = shallowMount(StaffAccountAdd, { 
      store, 
      localVue
    })
    it('has a name', () => {
      expect(StaffAccountAdd.name).toBe('StaffAccountAdd')
    })
    it('should render correct content', () => {
        expect(wrapper.find('#role-select').exists())
        expect(wrapper.find('#Add Dealerships').exists())
        expect(wrapper.find('#View Dealerships').exists())
        expect(wrapper.find('#Edit Dealerships').exists())
        expect(wrapper.find('#Add Staff Users').exists())
        expect(wrapper.find('#View Staff Users').exists())
        expect(wrapper.find('#Edit Staff Users').exists())
        expect(wrapper.find('#Delete Staff Users').exists())
        expect(wrapper.find('#Add Vehicles').exists())
        expect(wrapper.find('#View Vehicles').exists())
        expect(wrapper.find('#Edit Vehicles').exists())
        expect(wrapper.find('#Edit Vehicle Locations').exists())
        expect(wrapper.find('#Sell Vehicles').exists())
        expect(wrapper.find('#Delete Vehicles').exists())
        expect(wrapper.find('#Add Vehicle Properties').exists())
        expect(wrapper.find('#Edit Vehicle Properties').exists())
        expect(wrapper.find('#View Vehicle Properties').exists())
        expect(wrapper.find('#Delete Vehicle Properties').exists())
        expect(wrapper.find('#create-staff').exists())
    })
    test('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })