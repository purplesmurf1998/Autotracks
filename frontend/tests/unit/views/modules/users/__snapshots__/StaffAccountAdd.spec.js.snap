// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffAccountAdd.vue renders correctly 1`] = `
<div>
  <cform-stub>
    <!---->
     
    <crow-stub
      gutters="true"
      tag="div"
    >
      <ccol-stub
        tag="div"
      >
        <cinput-stub
          label="First Name"
          lazy="400"
          placeholder="User first name"
          type="text"
          value=""
        />
      </ccol-stub>
       
      <ccol-stub
        tag="div"
      >
        <cinput-stub
          label="Last Name"
          lazy="400"
          placeholder="User last name"
          type="text"
          value=""
        />
      </ccol-stub>
    </crow-stub>
     
    <cinput-stub
      label="Email"
      lazy="400"
      placeholder="User email address"
      type="text"
      value=""
    />
     
    <cselect-stub
      id="role-select"
      label="Account Role"
      options="Management,Sales Rep,Sales Rep + Showroom,Sales Rep + Demoline,Sales Rep + Benefits,Reception"
      placeholder="Please select a role"
      value=""
    />
     
    <label
      class="mb-0"
    >
      Account permissions
    </label>
     
    <hr />
     
    <crow-stub
      gutters="true"
      tag="div"
    >
      <ccol-stub
        tag="div"
      >
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Add Dealerships
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Add Dealerships"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              View Dealerships
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="View Dealerships"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Edit Dealerships
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Edit Dealerships"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Delete Dealerships
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Delete Dealerships"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Add Staff Users
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Add Staff Users"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              View Staff Users
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="View Staff Users"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Edit Staff Users
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Edit Staff Users"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Delete Staff Users
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Delete Staff Users"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Add Vehicles
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Add Vehicles"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
      </ccol-stub>
       
      <ccol-stub
        tag="div"
      >
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              View Vehicles
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="View Vehicles"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Edit Vehicles
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Edit Vehicles"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Edit Vehicle Locations
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Edit Vehicle Locations"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Sell Vehicles
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Sell Vehicles"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Delete Vehicles
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Delete Vehicles"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Add Vehicle Properties
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Add Vehicle Properties"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Edit Vehicle Properties
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Edit Vehicle Properties"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              View Vehicle Properties
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="View Vehicle Properties"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
        <crow-stub
          gutters="true"
          tag="div"
        >
          <ccol-stub
            tag="div"
          >
            <label>
              Delete Vehicle Properties
            </label>
          </ccol-stub>
           
          <ccol-stub
            class="d-flex justify-content-end"
            tag="div"
          >
            <cswitch-stub
              class="mr-1"
              color="success"
              id="Delete Vehicle Properties"
              shape="pill"
              type="checkbox"
            />
          </ccol-stub>
        </crow-stub>
      </ccol-stub>
    </crow-stub>
     
    <hr />
     
    <crow-stub
      gutters="true"
      tag="div"
    >
      <ccol-stub
        tag="div"
      >
        <label>
          Account Password
        </label>
      </ccol-stub>
       
      <ccol-stub
        class="d-flex align-items-end flex-column"
        tag="div"
      >
        <p
          style="color: red;"
        >
          1Z8F5HSEAMNO
        </p>
      </ccol-stub>
    </crow-stub>
     
    <label>
      Take note of the password and pass on to staff member. User will be
      prompted to change password upon first login.
    </label>
     
    <crow-stub
      class="justify-content-center mt-2"
      gutters="true"
      tag="div"
    >
      <cbutton-stub
        activeclass="router-link-active"
        color="primary"
        event="click"
        exactactiveclass="router-link-exact-active"
        id="create-staff"
        routertag="a"
        target="_self"
        type="submit"
      >
        
        Create
      
      </cbutton-stub>
       
      <cbutton-stub
        activeclass="router-link-active"
        class="ml-1"
        color="secondary"
        event="click"
        exactactiveclass="router-link-exact-active"
        routertag="a"
        target="_self"
        type="button"
      >
        
        Cancel
      
      </cbutton-stub>
    </crow-stub>
  </cform-stub>
</div>
`;
